internal/util/chanutil/debounced_chan_test.go

```diff
@@ -77,3 +80,63 @@ func TestDebouncedChan_OnlyBuffersOneEvent(t *testing.T) {
 	case <-time.After(20 * time.Millisecond):
 	}
 }
+
+func TestDebouncedChan_ContinuousOperation(t *testing.T) {
+	t.Parallel()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	const (
+		cooldown  = 17 * time.Millisecond
+		increment = 1 * time.Millisecond
+		testTime  = 150 * time.Millisecond
+	)
+
+	var (
+		debouncedChan = NewDebouncedChan(ctx, cooldown)
+		goroutineDone = make(chan struct{})
+		numSignals    int
+	)
+
+	go func() {
+		defer close(goroutineDone)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-debouncedChan.C():
+				numSignals++
+			}
+		}
+	}()
+
+	for tm := increment; tm <= testTime; tm += increment {
+		time.Sleep(increment)
+		debouncedChan.Call()
+	}
+
+	cancel()
+
+	select {
+	case <-goroutineDone:
+	case <-time.After(3 * time.Second):
+		require.FailNow(t, "Timed out waiting for goroutine to finish")
+	}
+
+	// Expect number of signals equal to number of cooldown periods that fit
+	// into our total test time, multiplied by two, because the debounced chan
+	// fires at the beginning and end of a bounce period. +1 for the last period
+	// that fires on the leading edge, but which we don't give time for the
+	// timer to fully expire. (We've chosen numbers so that test time doesn't
+	// divide by cooldown perfectly.)
```

Comment:
@bgentry Something I noticed while looking into the implementation is that for every cooldown period, given relatively constant `Call` signaling throughout, the debounce channel will actually fire twice for every period — once at the beginning, then once at the end as the timer fires. Assuming the signaling continues, it then fires immediately again. So the number of times it fires for N periods is actually 2*N.

It's not too bad of a problem, but OOC was that on purpose? I think there might be an alternate version of this that's more of a "rolling" debounce channel that fires once immediately, but then only fires again once at the end of every cooldown period.

Reply:
Nope, doesn't sound intentional at all. I think the purpose was for it to be a leading edge debounce algorithm (fire immediately, then do not fire again until (a) another signal comes in, and (b) at least `cooldown` has elapsed).